File: backend/runtime/internal/utils/snowflake/snowflake_test.go
package snowflake

import (
	"testing"
	"time"
)

func TestSnowflake(t *testing.T) {
	node, err := NewNode(1)
	if err != nil {
		t.Fatalf("Failed to create node: %v", err)
	}

	id1 := node.Generate()
	id2 := node.Generate()

	if id1 == id2 {
		t.Errorf("Generated duplicate IDs: %d", id1)
	}

	if id1 > id2 {
		t.Errorf("IDs are not monotonically increasing: %d > %d", id1, id2)
	}
}

func TestSnowflakeConcurrency(t *testing.T) {
	node, err := NewNode(1)
	if err != nil {
		t.Fatalf("Failed to create node: %v", err)
	}

	const numGoroutines = 100
	const numIDsPerGoroutine = 1000

	ids := make(chan int64, numGoroutines*numIDsPerGoroutine)

	for i := 0; i < numGoroutines; i++ {
		go func() {
			for j := 0; j < numIDsPerGoroutine; j++ {
				ids <- node.Generate()
			}
		}()
	}

	time.Sleep(100 * time.Millisecond)

	idMap := make(map[int64]bool)
	for i := 0; i < numGoroutines*numIDsPerGoroutine; i++ {
		id := <-ids
		if idMap[id] {
			t.Errorf("Duplicate ID found: %d", id)
		}
		idMap[id] = true
	}
}

func TestGlobalNode(t *testing.T) {
	err := InitGlobalNode(2)
	if err != nil {
		t.Fatalf("Failed to init global node: %v", err)
	}

	id1 := GenerateID()
	id2 := GenerateID()

	if id1 == id2 {
		t.Errorf("Generated duplicate IDs: %d", id1)
	}

	strID := GenerateStringID()
	if strID == "" {
		t.Errorf("Generated empty string ID")
	}
}

File: tests/frontend/src/shared/utils/snowflake.test.ts
import { describe, expect, it } from 'vitest';
import { Snowflake, generateId, initGlobalNode } from '../../../../../frontend/src/shared/utils/snowflake';

describe('Snowflake', () => {
  it('should generate unique IDs', () => {
    const node = new Snowflake(1);
    const id1 = node.generate();
    const id2 = node.generate();

    expect(id1).not.toBe(id2);
  });

  it('should generate monotonically increasing IDs', () => {
    const node = new Snowflake(1);
    const id1 = BigInt(node.generate());
    const id2 = BigInt(node.generate());

    expect(id1 < id2).toBe(true);
  });

  it('should throw error for invalid node ID', () => {
    expect(() => new Snowflake(-1)).toThrow();
    expect(() => new Snowflake(1024)).toThrow();
  });

  it('should generate IDs using global node', () => {
    initGlobalNode(2);
    const id1 = generateId();
    const id2 = generateId();

    expect(id1).not.toBe(id2);
    expect(typeof id1).toBe('string');
    expect(id1.length).toBeGreaterThan(0);
  });
});
